lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+            raise ValueError("Tolerances must be non-negative.")
+
+        if isinstance(self.atol, (int, float)) and isinstance(self.btol, (int, float)):
+            if self.atol == 0 and self.btol == 0 and self.max_steps is None:
+                raise ValueError(
+                    "Must specify `atol`, `btol`, or `max_steps` (or some combination "
+                    "of all three)."
+                )
+
+    def init(self, operator: AbstractLinearOperator, options: dict[str, Any]):
+        return operator
+
+    def compute(
+        self,
+        state: _LSMRState,
+        vector: PyTree[Array],
+        options: dict[str, Any],
+    ) -> tuple[PyTree[Array], RESULTS, dict[str, Any]]:
+        operator = state
+        x = options.get("y0", None)
+        damp = options.get("damp", 0.0)
+
+        m, n = operator.out_size(), operator.in_size()
+        # number of singular values
+        minDim = min([m, n])
+        if self.max_steps is None:
+            max_steps = minDim * 10  # for consistency with other iterative solvers
+        else:
+            max_steps = self.max_steps
+
+        if x is None:
+            x = jtu.tree_map(jnp.zeros_like, operator.in_structure())
+
+        b = vector
+        u = (ω(b) - ω(operator.mv(x))).ω
+        normb = self.norm(b)
+        beta = self.norm(u)
+
+        def beta_gt0(beta, u):
+            u = (ω(u) * (1 / beta)).ω
+            v = operator.T.mv(u)
+            alpha = self.norm(v)
+            return u, v, alpha
+
+        def beta_le0(beta, u):
+            v = jtu.tree_map(jnp.zeros_like, operator.in_structure())
+            alpha = 0.0
+            return u, v, alpha
+
+        u, v, alpha = lax.cond(beta > 0, beta_gt0, beta_le0, beta, u)
+        v = lax.cond(alpha > 0, lambda: (ω(v) / alpha).ω, lambda: v)
+
+        h = v.copy()
+        hbar = jtu.tree_map(jnp.zeros_like, operator.in_structure())
+
+        loop_state_vecs = (x, u, v, h, hbar)
+
+        # Initialize variables for 1st iteration.
+        itn = 0
+        zetabar = alpha * beta
+        alphabar = alpha
+        rho = 1.0
+        rhobar = 1.0
+        cbar = 1.0
+        sbar = 0.0
+
+        loop_state_main = (itn, alpha, beta, zetabar, alphabar, rho, rhobar, cbar, sbar)
+
+        # Initialize variables for estimation of ||r||.
+        betadd = beta
+        betad = 0.0
+        rhodold = 1.0
+        tautildeold = 0.0
+        thetatilde = 0.0
+        zeta = 0.0
+        d = 0.0
+
+        loop_state_r_est = (betadd, betad, rhodold, tautildeold, thetatilde, zeta, d)
+
+        # Initialize variables for estimation of ||A|| and cond(A)
+        normA2 = alpha**2
+        maxrbar = 0.0
+        minrbar = 1e100
+        normA = jnp.sqrt(normA2)
+        condA = 1.0
+
+        loop_state_anorm = (normA2, maxrbar, minrbar, normA, condA)
+
+        # Items for use in stopping rules, normb set earlier
+        istop = 0
+        ctol = 0.0
+        if self.conlim > 0:
+            ctol = 1 / self.conlim
```

Comment:
I'd prefer single-assignment form, i.e.
```python
if self.conlim > 0:
    ctol = 1 / self.conlim
else:
    ctol = 0
```

Reply:
Also perhaps this should be a `jnp.where`, just to support the (very edge) case that `self.conlim` is a traced value?